ee/src/main/java/org/jboss/as/ee/structure/InitializeInOrderProcessor.java

```diff
@@ -49,24 +50,45 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
-                for (final DeploymentUnit subDeployment : parent.getAttachment(org.jboss.as.server.deployment.Attachments.SUB_DEPLOYMENTS)) {
+        if (earConfig == null) {
+            return;
+        }
 
-                    final ResourceRoot deploymentRoot = subDeployment.getAttachment(org.jboss.as.server.deployment.Attachments.DEPLOYMENT_ROOT);
-                    final ModuleMetaData moduleMetaData = deploymentRoot.getAttachment(Attachments.MODULE_META_DATA);
-                    if (moduleMetaData != null) {
-                        deploymentUnitMap.put(moduleMetaData.getFileName(), subDeployment);
-                    }
+        final List<String> startupBeanModules = parent.getAttachmentList(org.jboss.as.ee.component.Attachments.DEPLOYMENT_STARTUP_MODULES);
+        final Map<String, DeploymentUnit> deploymentUnitMap = new HashMap<String, DeploymentUnit>();
+        final ResourceRoot deploymentRoot = deploymentUnit.getAttachment(org.jboss.as.server.deployment.Attachments.DEPLOYMENT_ROOT);
+        final ModuleMetaData thisModulesMetadata = deploymentRoot.getAttachment(Attachments.MODULE_META_DATA);
+
+        if((earConfig.getInitializeInOrder() || (!startupBeanModules.isEmpty())) && !startupBeanModules.contains(deploymentUnit.getName())) {
+            for (final DeploymentUnit subDeployment : parent.getAttachment(org.jboss.as.server.deployment.Attachments.SUB_DEPLOYMENTS)) {
+
+                final ResourceRoot subdeploymentRoot = subDeployment.getAttachment(org.jboss.as.server.deployment.Attachments.DEPLOYMENT_ROOT);
+                final ModuleMetaData moduleMetaData = subdeploymentRoot.getAttachment(Attachments.MODULE_META_DATA);
+                if (moduleMetaData != null) {
+                    deploymentUnitMap.put(moduleMetaData.getFileName(), subDeployment);
+                }
+            }
+        }
+
+        if(!startupBeanModules.isEmpty() && earConfig.getModules().size() > 1 && !startupBeanModules.contains(deploymentUnit.getName())) {
+            if (thisModulesMetadata != null && thisModulesMetadata.getType() != ModuleMetaData.ModuleType.Client) {
+                for (String startupModule : startupBeanModules) {
+                    final ServiceName serviceName = Services.deploymentUnitName(parent.getName(), startupModule);
+                    phaseContext.addToAttachmentList(org.jboss.as.server.deployment.Attachments.NEXT_PHASE_DEPS, serviceName.append(Phase.INSTALL.name()));
+                    final DeploymentUnit prevDeployment = deploymentUnitMap.get(startupModule);
+
+                    phaseContext.addToAttachmentList(org.jboss.as.server.deployment.Attachments.NEXT_PHASE_DEPS, DeploymentCompleteServiceProcessor.serviceName(prevDeployment.getServiceName()));
                 }
+            }
+        }
 
+        if (!startupBeanModules.contains(deploymentUnit.getName())) {
+            final boolean inOrder = earConfig.getInitializeInOrder();
```

Comment:
`inOrder` is used multiple times so we can move the assignment line up and use `inOrder` variables where needed to make it more readable.